std/typecons.d

```diff
@@ -4865,12 +4865,154 @@ unittest
+        @property nothrow @safe
+        bool isInitialized() const
+        {
+            return _store !is null;
+        }
+
+        /// Get the underlying reference count.
+        @property nothrow @safe
+        size_t refCount() const
+        {
+            return _store._count;
+        }
+
+    private:
+
+        void destroy()
+        {
+            import core.memory : GC;
+            import core.stdc.stdlib : free;
+
+            _store._destroy(cast(T)_store._payload.ptr);
+            static if (hasIndirections!T)
+                GC.removeRange(_store._payload.ptr);
+            free(_store);
+            _store = null;
+        }
+    }
+    private RefCountedStore _refCounted;
+
+    /// Returns storage implementation struct.
+    @property nothrow @safe
+    ref inout(RefCountedStore) refCountedStore() inout
+    {
+        return _refCounted;
+    }
+
+    /// Boolean conversion for RefCounted, true when initialized.
+    bool opCast(T:bool)() const @safe
+    {
+        return _refCounted.isInitialized;
+    }
+
+    /// copy constructor for other RefCounted class
+    this(U)(RefCounted!U u)
+        if (is(U == class) && is(U : T))
+    {
+        // possibly polymorph conversion
+        _refCounted._store = cast(RefCountedStore.Store*)u._refCounted._store;
+        // incref
+        if (_refCounted._store !is null)
+            ++_refCounted._store._count;
+    }
+
+    /// Postblit constructor to track reference count.
+    this(this)
+    {
+        ++_refCounted._store._count;
+    }
+
+    /**
+       Destructor that destroys and frees the payload when the ref count goes to zero.
+
+       Note: This doesn't destroy a possibly attached monitor and won't call dispose events.
+    */
+    ~this()
+    {
+        if (_refCounted._store is null || --_refCounted._store._count)
+            return;
+        _refCounted.destroy();
+    }
+
+    /// assign another ref counted class (can be a derived class)
+    void opAssign(U)(RefCounted!U u)
+        if (is(U == class) && is(U : T))
+    {
+        if (_refCounted._store !is null)
+        {
+            if (--_refCounted._store._count == 0)
+                _refCounted.destroy();
+        }
+        // possibly polymorph conversion
+        _refCounted._store = cast(RefCountedStore.Store*)u._refCounted._store;
+        // incref
+        if (_refCounted._store !is null)
+            ++_refCounted._store._count;
+    }
+
+    /// nullify this RefCounted
+    void opAssign(typeof(null))
+    {
+        if (_refCounted._store !is null)
+        {
+            if (--_refCounted._store._count == 0)
+                _refCounted.destroy();
+        }
+        _refCounted._store = null;
+    }
+
+    /**
+       Get the underlying payload.
+
+       Note that getting a class reference is currently unsafe
+       as there is currently no way to stop it from escaping. (see DIP69)
+     */
+    @property nothrow @system
```

Comment:
Any reason it's explicitly marked `nothrow` but relies on attribute inference for `pure` and `@nogc`?